### ISSUE TITLE
Is there any way to enable "async" mode like nfs kernel server?

### ISSUE THREAD

**Issue by kaoet** [nfs-ganesha/nfs-ganesha#218]
With kernel server, I can write `*(async,...)` in `/etc/exports` which results in extremely fast operations.

I don't care lost of my data. Is it possible to enable such feature in nfs-ganesha?

**Comment by guihecheng** [comment]
AFAIK, nfs-ganesha supports sync mode operations only, but there are former calls for aync APIs, so I think there shall be some undergoing works, but not revealed yet, thanks :)

**Comment by ffilz** [comment]
We do not support knfsd's mis-named async option. That option bypasses assurances of synced data even when clients request synchronous writes. Sure, it makes for very fast response times, but at the expense of any assurances of data integrity.

There is some work underway to allow Ganesha to make async FSAL calls. These would still assure sync to disk as instructed by the client, but would free Ganesha's threads from blocking on I/O. That will improve Ganesha thread utilization, and probably improve throughput somewhat, however, a client COMMIT operation or a client sync WRITE would still delay response to the client until the data was committed to disk.

What WOULD improve response time would be underlying storage that had fast commit to NVRAM with write back to permanent storage. This would allow the storage to report completion of write I/O faster and this allow Ganesha to respond to a COMMIT faster.

**Comment by renhwztetecs** [comment]
I am very interested in "async FSAL calls", because in asynchronous mode, concurrent single-write scenarios are based on certain advantages. I don't know when it will be available? Or if there is a BP or the like, we can participate in it.

**Comment by dang** [comment]
That was merged on Friday, so it's in now.

**Comment by ffilz** [comment]
Note that the async that was merged Friday is NOT related to the kernel async export option.

What was merged Friday also doesn't actually do anything yet. Now that it's present, it will require FSAL (and likely underlying filesystem) work to actually implement async. There is test/example code in FSAL_MEM but of course FSAL_MEM doesn't really have to wait on any I/O so it doesn't do much other than provide a platform to test the upper layer code.

Further efforts here should actually be discussed in a new issue or better yet, on the Devel mailing list on http://lists.nfs-ganesha.org/

Frank

**Comment by ffilz** [comment]
Closing as resolved.
